### PR TITLE
Support scaling for benchmark data with flexible call to sklearn

### PR DESCRIPTION
**Pull request by qzheng75**
Changes in this PR:
1. In `wod_predictor/feature_engineering_parts/benchmark_stats_fe.py`, the initialization of the scaler is modified to support the GenericSklearnScaler. Related data scaling for benchmark stats parts are implemented.
2. In `wod_predictor/feature_engineering_parts/helpers.py`, add imputing with mean and median.
3. In `wod_predictor/feature_engineering_parts/normalization.py`, implemented the GenericSklearnScaler class with documentation about usage. It can be used to scale both the open_results and benchmark_stats. The pipeline notebook is reflected to fit the current scaling pipeline.
4. Provided 8 unit tests for the implemented GenericSklearnScaler

### CODE REVIEW

File: tests/test_generic_sklearn_scaler.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        transformed = scaler.transform(single_row)
+        self.assertEqual(len(transformed), 1)
+
+
+if __name__ == '__main__':
+    unittest.main()
+    # suite = unittest.TestSuite()
+    # suite.addTest(TestGenericSklearnScaler('test_edge_cases'))    
+    # runner = unittest.TextTestRunner()
+    # runner.run(suite)
```
Comment (hassannaveed1997) [inline tests/test_generic_sklearn_scaler.py:142]: Can you delete this dead code?

File: tests/test_generic_sklearn_scaler.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        single_col = self.numeric_data[['Back Squat']]
+        scaler.fit(single_col)
+        transformed = scaler.transform(single_col)
+        self.assertEqual(transformed.shape[1], 1)
+
+        # single row
+        single_row = self.numeric_data.iloc[[0]]
+        scaler.fit(single_row)
+        transformed = scaler.transform(single_row)
+        self.assertEqual(len(transformed), 1)
```
[the remark this answers is not in the document]
Comment (hassannaveed1997) [reply tests/test_generic_sklearn_scaler.py:134]: Another edge case can be non_numeric values in columns

File: examples/example_pipeline.ipynb [hunk 3]
```diff
@@ -36,21 +47,38 @@
...
     "preprocessing_config = {\n",
     "    'open_results':\n",
     "    {\n",
     "        'scale_up': True,\n",
-    "        'scale_method': 'standard', # standard or quantile\n",
-    "\n",
+    "        'scale_args': {\n",
+    "            \"method\": \"general\",\n",
+    "            \"scaler_name\": \"StandardScaler\"\n",
+    "        }\n",
```
Comment (hassannaveed1997) [inline examples/example_pipeline.ipynb:72]: Good work adding the new options in the configs.

File: wod_predictor/feature_engineering_parts/benchmark_stats_fe.py [hunk 4]
```diff
@@ -33,16 +42,41 @@ def __init__(
...
+                self.scaler = None
 
     def fit(self, benchmark_data):
         """
         Initialize any transformers for later use in transform
         """
+        df_copy = benchmark_data.copy()
+
+        # TODO:
+        # These redundant fits need to be addressed in a future PR.
```
Comment (hassannaveed1997) [inline wod_predictor/feature_engineering_parts/benchmark_stats_fe.py:66]: Redundant transforms?

File: wod_predictor/feature_engineering_parts/normalization.py [hunk 5]
```diff
@@ -1,24 +1,37 @@
...
 
+    def pad_columns(self, df, full_cols=None):
+        """Adds missing columns to prevent raising error"""
+        if full_cols is None:
+            full_cols = self.scaler.feature_names_in_
+        for col in full_cols:
+            if col not in df.columns:
+                df[col] = np.nan
+        return df
+
```
Comment (hassannaveed1997) [inline wod_predictor/feature_engineering_parts/normalization.py:27]: Now I understand how you fixed the issue I was having. Good work 👏

**Review by hassannaveed1997 — CHANGES_REQUESTED** [review]
This is great work, really impressed by how you implemented the general scaler class. Can tell we have a real SWE here 🔥 
Left some comments, most of them are pretty minor, and some of them don't need action.

File: examples/example_pipeline.ipynb [hunk 6]
```diff
@@ -10,14 +10,25 @@
...
+    "from wod_predictor.modeling import RandomForestModel"
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 2,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
```
[the remark this answers is not in the document]
Comment (qzheng75) [reply examples/example_pipeline.ipynb]: Should be addressed in the latest commit.

File: tests/test_generic_sklearn_scaler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (qzheng75) [reply tests/test_generic_sklearn_scaler.py]: Seperated them into 3 test cases in the latest commit.

File: tests/test_generic_sklearn_scaler.py (hunk 1 above)
Comment (qzheng75) [reply tests/test_generic_sklearn_scaler.py]: I deleted the commented code, but keep `unittest.main()` so that the tests can be ran.

File: wod_predictor/feature_engineering_parts/benchmark_stats_fe.py (hunk 4 above)
Comment (qzheng75) [reply wod_predictor/feature_engineering_parts/benchmark_stats_fe.py]: Refined comments in the latest commit.

**Review by hassannaveed1997 — APPROVED** [review]
This is great work, thanks for also adding unit tests. Gave it a run and worked fine.
